### PR TITLE
XPU: Support head_dim=80 via padding for  flash attention SYCL kernel

### PR DESCRIPTION
**Pull request by kpjeeja**
The SYCL flash attention kernel in sgl-kernel-xpu only supports specific head dimensions: 64, 96, 128, 192. Models like Qwen2-VL use head_dim=80 in their vision encoders (1280 hidden_dim / 16 heads = 80), which causes runtime failures when using XPU flash attention.

This patch implements a padding workaround in VisionIntelXPUAttention:
1. Check if head_dim is in the supported list [64, 96, 128, 192]
2. If unsupported, pad q/k/v to the nearest larger supported size
3. Run flash attention with padded tensors
4. Slice output back to original head_dim

Example: head_dim=80 → pad to 96 → compute attention → slice to 80

This maintains correctness while enabling XPU flash attention for models with non-standard head dimensions.

Tested with: Qwen/Qwen2-VL-2B-Instruct (vision encoder head_dim=80)

---
### CI States

Latest PR Test (Base): :x: [Run #27751359404](https://github.com/sgl-project/sglang/actions/runs/27751359404)
Latest PR Test (Extra): :x: [Run #27751359132](https://github.com/sgl-project/sglang/actions/runs/27751359132)

### CODE REVIEW

File: test/registered/xpu/test_xpu_head_dim_padding.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        # Verify output shape
+        self.assertEqual(
+            output_80.shape,
+            (total_tokens, num_heads, head_dim),
+            "Output should maintain original head_dim=80",
+        )
+
+        # Verify reasonable values (not all zeros, no NaN/Inf)
+        self.assertTrue(torch.isfinite(output_80).all())
+        self.assertGreater(output_80.abs().mean().item(), 0)
```
[the remark this answers is not in the document]
Comment (kpjeeja) [reply test/registered/xpu/test_xpu_head_dim_padding.py]: done

File: python/sglang/srt/layers/attention/vision.py [hunk 2]
```diff
@@ -854,7 +875,11 @@ def forward(
         )
         if s_aux is not None:
             fa_kwargs["sinks"] = s_aux
-        output = flash_attn_varlen_func(q, k, v, **fa_kwargs)
+        output = flash_attn_varlen_func(q_padded, k_padded, v_padded, **fa_kwargs)
```
[the remark this answers is not in the document]
Comment (kpjeeja) [reply python/sglang/srt/layers/attention/vision.py:888]: fixed

File: python/sglang/srt/layers/attention/vision.py [hunk 3]
```diff
@@ -844,6 +844,37 @@ def forward(
...
 
+        # XPU flash attention supports head sizes: 64, 96, 128, 192
+        # Pad unsupported head sizes to the nearest supported size
+        head_size = q.size(-1)
+        supported_head_sizes = [64, 96, 128, 192]
+
+        # Compute scale from original head_size (before padding)
+        # Attention scaling should reflect the actual embedding dimension
+        if softmax_scale is None:
+            softmax_scale = head_size**-0.5
```
Comment (mickqian) [inline python/sglang/srt/layers/attention/vision.py:855]: please move this and `supported_head_sizes` out of forward hot path

**Comment by mingfeima** [comment]
there are a couple of issues in this PR:

* generally, adding check logic in forward is not a good idea. This would add a heavy burden in each forward run, especially xpu kernel launch has bigger overhead compared v.s. cuda. and this patch uses padding for unsupported cases, and it is better to simply report a reasonable error from sgl-kernel-xpu (which i believed already done). it is better than a silent workaround with bad performance. and for the unsupported cases, we later fix it inside sgl-kernel-xpu.
* test cases. now we put kernel level test inside sgl-kernel-xpu, on sglang we usually run model e2e test. (some sglang xpu test cases are not following this rule, we will remove them later on).

more generally, it should be better to determine which backend to use when init the model or parsing the argument, but not inside `forward`. but as i commented above, i prefer `human readable error + quick fix` over `silent work around` (this would result into perf issue which may lurk for a long time if not paying special attention to.

**Comment by mingfeima** [comment]
cc @jianan-gu

File: test/registered/xpu/test_xpu_head_dim_padding.py [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+        )
+        v = torch.randn(
+            total_tokens, num_heads, head_dim, dtype=self.dtype, device=self.device
+        )
+
+        cu_seqlens = torch.tensor([0, seq_len], dtype=torch.int32, device=self.device)
+
+        # Run with XPU attention (automatic padding head_dim=80 -> 96)
+        xpu_attn = VisionIntelXPUAttention(
+            num_heads=num_heads,
```
Comment (siju-samuel) [inline test/registered/xpu/test_xpu_head_dim_padding.py:180]: scale=, head_size=, num_kv_heads= to the [VisionIntelXPUAttention constructor, but its __init__](https://github.com/sgl-project/sglang/blob/47ae1241d3ebf943cf5ca447064030ef1425ba28/python/sglang/srt/layers/attention/vision.py#L813) takes only **kwargs and ignores them

File: test/registered/xpu/test_xpu_head_dim_padding.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+        output_std = output_xpu.std().item()
+        self.assertGreater(output_std, 0.1, "Output std should be > 0.1")
+        self.assertLess(output_std, 10.0, "Output std should be < 10.0")
+
+        # Verify different runs with same input produce same output (deterministic)
+        output_xpu2 = xpu_attn(
+            q, k, v, cu_seqlens=cu_seqlens, bsz=batch_size, seq_len=seq_len
+        )
+        torch.testing.assert_close(
+            output_xpu, output_xpu2, msg="XPU attention should be deterministic"
```
Comment (siju-samuel) [inline test/registered/xpu/test_xpu_head_dim_padding.py:213]: i think here its checked only for determinism, but need to evaluate for corectness as well.

**Comment by kpjeeja** [comment]
> there are a couple of issues in this PR:
> 
> * generally, adding check logic in forward is not a good idea. This would add a heavy burden in each forward run, especially xpu kernel launch has bigger overhead compared v.s. cuda. and this patch uses padding for unsupported cases, and it is better to simply report a reasonable error from sgl-kernel-xpu (which i believed already done). it is better than a silent workaround with bad performance. and for the unsupported cases, we later fix it inside sgl-kernel-xpu.
> * test cases. now we put kernel level test inside sgl-kernel-xpu, on sglang we usually run model e2e test. (some sglang xpu test cases are not following this rule, we will remove them later on).
> 
> more generally, it should be better to determine which backend to use when init the model or parsing the argument, but not inside `forward`. but as i commented above, i prefer `human readable error + quick fix` over `silent work around` (this would result into perf issue which may lurk for a long time if not paying special attention to.

yes, i agree already the sgl-kernel provides the correct error msg. i will move this to sgl-kernel-xpu implementation to support this case
